### DISCUSSION TITLE
After 0.22, how do I load a .so language?

### DISCUSSION THREAD

**Discussion by realhackcraft** [tree-sitter/py-tree-sitter#251] [Q&A] [answered]
For some reason, pip install tree-sitter-python (I need to parse python code in python) stopped working. I managed to compile the python TS language from source. Now, I tried to load it, but since 0.22 removed `Language(path, name)`, how can I load the `.so` binary?

**Comment by ObserverOfTime** [comment]
```
pip install git+https://github.com/tree-sitter/tree-sitter-python
```

**Reply by Moosems** [reply]
What if its some random tree sitter that doesn't have a pypi release?

**Reply by realhackcraft** [reply]
Exactly. Or if I needed to make my own, unofficial grammar?

**Reply by jdugan6240** [reply, marked as answer]
Basically, you'd need to manually do what py-tree-sitter used to do itself - that is, grabbing a pointer to the `tree_sitter_python` function from the .so library file and passing that pointer to the Language constructor.

Something like the following should work:

```python
from tree_sitter import Language, Parser
from ctypes import cdll, c_void_p
from os import fspath

def lang_from_so(path: str, name: str) -> Language:
    lib = cdll.LoadLibrary(fspath(path))
    language_function = getattr(lib, f"tree_sitter_{name}")
    language_function.restype = c_void_p
    language_ptr = language_function()
    return Language(language_ptr)

PY_LANG = lang_from_so(path_to_python_so, "python")
PY_PARSER = Parser()
PY_PARSER.set_language(PY_LANG)

# Do whatever with language and parser objects....
```

**Reply by Moosems** [reply]
@jdugan6240 Thanks :)
